### PR TITLE
[WIP] Refine MultiDevSSAGraph

### PR DESCRIPTION
**Pull request by chengduoZH**
This PR removes the strict checking of Op role.
  - Remove `CheckProgram`
  - Remove `SortOpsAndDelayOptimizeOp`

### CODE REVIEW

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 1]
```diff
@@ -303,32 +181,23 @@ std::unique_ptr<ir::Graph> MultiDevSSAGraphBuilder::ApplyImpl(
...
 
-  std::unordered_map<std::string, int> sharded_var_device;
+  std::unordered_map<std::string, int> shared_var_device;
 
   for (ir::Node *node : sorted_ops) {
-    if (boost::get<int>(
-            node->Op()->GetAttr(OpProtoAndCheckerMaker::OpRoleAttrName())) ==
-        static_cast<int>(OpRole::kRPC)) {
-      int op_dev_id = CreateRPCOp(&result, node, &sharded_var_device);
+    if (IsSameOpRole(*node, OpRole::kRPC)) {
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: IsSameOpRole => OpHaveRole

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 2]
```diff
@@ -303,32 +181,23 @@ std::unique_ptr<ir::Graph> MultiDevSSAGraphBuilder::ApplyImpl(
...
-
   std::vector<std::unordered_set<std::string>> bcast_var_name_set;
   bcast_var_name_set.resize(places_.size());
 
-  size_t cur_device_id = 0;
   bool is_forwarding = true;
   bool is_dist_train = false;
 
-  std::unordered_map<std::string, int> sharded_var_device;
+  std::unordered_map<std::string, int> shared_var_device;
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: sharded means variables are spread to different devices, should not change the word meaning.

**Review by typhoonzero — APPROVED** [review]
LGTM

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc]: Done

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc]: Done

**Pull request merged**